### PR TITLE
Add forward AD support for miscellanous operators

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67045
* #67041
* __->__ #67040
* #66294
* #66293
* #66292

 - Add forward AD formulas for:
   - Public API `relu`, `stack`, `neg`, `binary_cross_entropy_with_logits`
   - Non-public-API: `threshold_backward`
 - We rely on OpInfos to test all them except `binary_cross_entropy_with_logits`.
 - Temporary: applies a fix for non-differentiable ops, but will be rebased away later

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1642,6 +1644,7 @@
 
 - name: relu(Tensor self) -> Tensor
   self: threshold_backward(grad, result, 0)
+  result: threshold_backward(self_t, self_p, 0)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1647]: auto_element_wise

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -1055,6 +1055,7 @@
 
 - name: neg(Tensor self) -> Tensor
   self: grad.neg()
+  result: self_t.neg()
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1058]: auto_element_wise

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -2053,6 +2056,7 @@
 - name: threshold_backward(Tensor grad_output, Tensor self, Scalar threshold) -> Tensor
   grad_output: threshold_backward(grad, self, threshold)
   self: zeros_like(grad)
+  result: zeros_like(self_t) + threshold_backward(grad_output_t, self_p, threshold)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:2059]: What is the zeros_like for here?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -1332,6 +1342,37 @@ Tensor binary_cross_entropy_with_logits_target_backward(const Tensor& grad_outpu
...
+  const Tensor& weight = *weight_maybe_owned;
+  const Tensor& pos_weight = c10::value_or_else(pos_weight_opt, [] {return Tensor();});
+
+  Tensor grad_input;
+  Tensor grad_target;
+
+  if (pos_weight.defined()) {
+    // pos_weight need to be broadcasted, thus mul(target) is not inplace.
+    auto t = pos_weight.mul(target_p);
+    grad_input = input_t.mul_(t.add(1).sub_(target_p).mul_(input_p.sigmoid()).sub_(t));
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1357]: The tangent should not be modified in the forward was not inplace.

File: tools/autograd/derivatives.yaml (hunk 3 above)
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:2059]: the one in the native namespace?

Comment (albanD) [reply tools/autograd/derivatives.yaml:2074]: Ho I meant why are you adding 0s to something? Why not just have the something?

**Review by albanD — APPROVED** [review]
LGTM as long as the CI is green

File: tools/autograd/derivatives.yaml (hunk 3 above)
Comment (soulitzer) [reply tools/autograd/derivatives.yaml:2074]: Ah it should be for when the self has tangent but grad_output does not, then LHS is batched but RHS is not, so we'd like to preserve the batchedness.

**Comment by soulitzer** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/67820

**Pull request closed without merging**